herddb-core/src/main/java/herddb/client/HDBConnection.java

```diff
@@ -186,14 +194,83 @@ public DMLResult executeUpdate(String tableSpace, String query, long tx, boolean
                 return route.executeUpdate(tableSpace, query, tx, returnValues, usePreparedStatement, params);
             } catch (RetryRequestException retry) {
                 LOGGER.log(Level.SEVERE, "error " + retry, retry);
-                sleepOnRetry();
+                handleRetryError(retry);
             }
         }
         throw new HDBException("client is closed");
     }
 
+    public CompletableFuture<DMLResult> executeUpdateAsync(String tableSpace, String query, long tx, boolean returnValues, boolean usePreparedStatement, List<Object> params) {
+        if (discoverTablespaceFromSql) {
+            tableSpace = discoverTablespace(tableSpace, query);
+        }
+        if (closed) {
+            return FutureUtils.exception(new HDBException("client is closed"));
+        }
+        CompletableFuture<DMLResult> res = new CompletableFuture<>();
+
+        AtomicInteger count = new AtomicInteger(10);
+        executeStatementAsyncInternal(tableSpace, res, query, tx, returnValues, usePreparedStatement, params, count);
+        return res;
+    }
+
+    private void executeStatementAsyncInternal(String tableSpace, CompletableFuture<DMLResult> res, String query, long tx, boolean returnValues, boolean usePreparedStatement, List<Object> params, AtomicInteger count) {
```

Comment:
Wouldn't be cleaner to return a CompletableFuture and not require it as an input?

Reply:
No, it is like a recursion...